오픈마켓 [Step 1] Gundy, Wonbi

wonbi92 opened the pull request:
@wonhee009
안녕하세요 라자냐~ 첫 번째 PR을 보내게 되어서 기분이 좋자냐~
여러부분 신경써서 구현하고자 하였습니다!
부족한 부분이나 잘못된 부분이 있다면 가차없는 피드백 부탁드립니다!!
잘 부탁 드립니다!!

## ✓ 진행단계 : Step-1

## 📝 STEP 진행 중 경험하고 배운 것
- JSONParsing
    ☑️ DTO 생성
- Networking 구현
    ☑️ URLSession을 활용한 서버와의 데이터 통신
    ☑️ 각 네트워킹 요소를 프로토콜을 이용하여 추상화
- Test Double 작성
    ☑️ 테스트를 위한 객체 생성
    ☑️ Unit Test 진행

## ⚒️ 코드 구현내용 
<details>
<summary>구현내용</summary>
    
###  구현
- Network
    - HttpMethod 
        - HttpMethod를 나타내는 열거형 타입입니다.
    - NetworkRequest
        - 네트워킹을 위한 URL과 Request를 가지고 이를 구현하기 위한 필수 프로퍼티를 선언하는 프로토콜입니다.
    - NetworkManager
        - 네트워크에서 데이터를 가져와 오류를 처리하고 데이터를 파싱해주는 객체입니다.
    - URLSessionProtocol
        - DIP적용을 위해 `dataTask`메서드를 정의하는 프로토콜입니다.
        - 이 프로토콜을 채택하면 `dataTask`메서드의 로직을 구현해주어야 합니다.
    - URLSessionDataTaskProtocol
        - `URLSessionProtocol`의 `dataTask`메서드에서 반환하는 타입을 지정하는 프로토콜입니다.
        - 이 프로토콜을 채택하면 `resume`메서드의 로직을 구현해주어야 합니다.
- Extension
    - JSONDecoder
        - 제네릭 타입과 데이터를 받아 디코딩하는 타입 메서드를 추가하였습니다.
    - String
        - `"yyyy-MM-dd'T'HH:mm:ss"`의 형식의 문자열을 `Date`타입의 값으로 변경시켜주는 메서드를 추가하였습니다.
- OpenMarket
    - Product
        - `Codable`을 채택하는 DTO입니다.
    - ProductList
        - `Codable`을 채택하는 DTO입니다.
    - HealthCheckerRequest
        - NetworkRequest를 채택하고, Application HealthChekcer를 리퀘스트하기위한 프로퍼티를 갖고 있는 구조체입니다.
    - ProductListRequest
        - NetworkRequest를 채택하고, 상품 리스트 조회를 리퀘스트하기위한 프로퍼티를 갖고 있는 구조체입니다.
    - ProductDetailRequest
        - NetworkRequest를 채택하고, 상품 상세 조회를 리퀘스트하기위한 프로퍼티를 갖고 있는 구조체입니다.
- OpenMarketTests
    - products
        - 테스트를 위한 Mock JSON데이터입니다.
    - DataLoader
        - Mock JSON데이터를 코드로 연결시켜 data를 생성해주는 클래스입니다.
    - DummyData
        - 실제 네트워킹이 아닌 테스트를 진행할 때 반환할 데이터 구조체입니다.
    - MockURLSession
        - 테스트를 위해 실제 네트워킹이 아니라 `DummyData`를 사용하는 클래스입니다.
    - MockURLSessionDataTask
        - 테스트를 위해 실제 네트워킹 테스트가 아닌 `DummyData`를 반환하는 클래스 입니다.
    
### Unit Test 작성
- JSONDecoder, DTO
    - JSONParsingTests
- NetworkManager
    - NetworkManagerTests
- OpenMarketNetworkRequest
    - NetworkRequestTests
    
</details>


## 💭 고민한 부분

### 테스트용 JSON 파일과 서버 API 문서
- 이번 프로젝트 안내페이지에서 제공하는 테스트용 JSON 파일과 앞으로 작업을 진행할 API의 JSON이 서로 CodingKey나 value의 형식이 달랐습니다. 
- 우선 테스트용 파일에 맞게 DTO를 구현했었는데, 실제 API Network를 진행해보면서 서로 다르다는 것을 알게 되었습니다. 
- 테스트의 취지가 네트워크가 없는 상황에서도 정상적으로 동작하는지를 검증하는 것이기 때문에 서버 API 문서의 데이터 형식에 맞추는 것이 적절하다고 생각해 테스트용 JSON 파일을 서버 API 문서의 데이터 형식에 맞게 수정하였습니다. 

### DIP적용을 위한 Protocol 구현하기
- 실제로 네트워킹을 하지 않고, 정상적으로 fetch가 진행되는지 로직을 테스트 하기위해 `MockURLSession`이라는 test double객체를 만들고 `MockData` 객체를 반환하도록 로직을 구현해야 했습니다.
- 이 과정에서 URLSession의 dataTask메서드에서 네트워킹 하는 부분을가로채 실제 네트워킹이 아니라 미리 만들어둔 `MockData` 객체를 반환하게 구현하여야 했습니다.
- 실제 `URLSession`과 이런 `MockURLSession`은 다르게 작동해야 하므로, 프로토콜을 통해 의존성을 역전시켜 네트워크 매니저가 프로토콜을 바라보게 하여 테스트를 할 때는 `MockURLSession`을 주입시키고, 실제 네트워킹을 할 때는 `URLSession`을 주입시키는 방법으로 진행해 보았습니다.
- 이 때 `URLSessionDataTask`타입도 Mock데이터로 만들어 주었는데, 이 타입의 초기화 구문이 iOS13버전 부터 더 이상 사용되지 않는(deprecated) 로직이어서 이 타입도 프로토콜을 채택하게 하여 구현 해 보았습니다.

## ❓ 조언을 얻고 싶은 부분
### 재사용이 가능한 Request객체 구현하기
- 이번 프로젝트에서 네트워킹의 요소가 총 3가지 였습니다.
- 그리고 실제 오픈마켓 서버에서도 GET메서드 말고도 POST, PATCH, DEL 등의 httpMethod를 사용할 수 있었습니다.
- 이에 각각의 리퀘스트마다 URL주소가 달라졌었고, 이를 String으로 써주는건 코드의 재사용이 많고, 확장성에도 문제가 있어보인다 판단하였습니다.
- 그래서 `NetworkRequest` 프로토콜을 만들고, 각 네트워킹 요소를 객체화하여 이를 채택하도록 하였습니다.
- 이로써 URL을 String으로 직접 작성할 필요가 없이 프로토콜을 채택하고 그 프로토콜을 각자의 역할에 맞게 구현만 해주면 알아서 URL과 Request를 만들게 되었습니다.
- 혹시 저희가 구현한 `NetworkRequest`에서 부족한 부분이나 잘못된 부분, 혹은 개선할 부분이 있을까요??

### 각 파일의 위치가 적절한지
- 테스트를 위한 JSON 파일은 `OpenMarketTests` 내부에 위치시켰고, 범용적으로 네트워크에 사용할 수 있는 프로토콜이나 타입은 `OpenMarket` 외부에 `Network`라는 그룹을 따로 생성하여 관리하였습니다. Network 기능을 수행하지만 그 용도가 OpenMarket 프로젝트에만 사용될 `OpenMarketNetworkRequest` 등은 `OpenMarket` 내부에 위치시켰습니다.
- 저희가 네트워크 통신 작업이 처음이다보니, 이런 방식으로 파일을 정리하는 것이 적절한지 조언을 얻고 싶습니다.

### BoringSSL에 대하여
- Step 1-2를 진행하면서 서버와 실제로 데이터를 주고 받을 때, `boringssl_metrics_log_metric_block_invoke(153)` 라는 메시지가 콘솔에 뜨게 되었습니다. 
- [boringssl_metrics_log_metric_block_invoke(151) Failed to log metrics](https://github.com/firebase/firebase-ios-sdk/issues/9262)의 내용 및 여러 게시글들을 찾아본 결과 기능에는 문제가 없고, 콘솔에 나타나기만 하는 메시지로 파악했습니다. 
- 스키마에서 OS_ACTIVITY_MODE를 disable로 바꿔주면 이 에러 메세지가 콘솔에서 나타나지 않지만, `NSLog`도 사라지는 문제가 있어서 이 방법은 적절한 방법이 아니라 판단하였습니다.
- 또 다른 방법으로는 `xcrun simctl spawn booted log config --subsystem com.apple.network --category boringssl --mode level:off` 처럼 Xcode 설정을 건드리는 방법도 있었는데, 이는 Xcode의 설정을 건드리는 것이고, 어떤 사이드 이펙트가 발생할지 몰라, 적용하기에 적절하지 않다 판단하여 적용하지 않았습니다.
- 혹시 이 메시지를 신경쓰지 않고 작업을 이어나가도 괜찮은지 여쭙고 싶습니다!

wonbi92 committed:
feat : 파싱할 데이터 DTO 생성

- asset에 Mock데이터 추가
- ProductList, Product, Currency 타입 생성

Gundy93 committed:
feat: json 파일을 파싱할 타입 메서드  생성

wonbi92 committed:
test : JSON디코딩 유닛 테스트 추가, DateFormatter 생성

- DTO 소스파일 분리
- String 익스텐션을 통해 Date 타입으로 변환해주는 DateFormatter 메서드 생성
- Product타입 내부 Date타입의 연산 프로퍼티 선언
- ProductList타입 파싱에 대한 테스트 케이스 작성
- Product타입 파싱 및 연산 프로퍼티에 대한 테스트 케이스 작성

Gundy93 committed:
test: 유닛 테스트 추가

wonbi92 committed:
feat : HttpMethod 열거형, NetworkRequest 프로토콜 생성

- NetworkRequest 프로토콜 및  익스텐션을 통해 urlComponents, requestURL 프로퍼티 기본 구현
- HttpMethod 열거형 생성

Gundy93 committed:
feat: NetworkRequest를 채택하는 타입 구현

- HealthCheckerRequest, ProductListRequest, ProductDetailRequest 구현

wonbi92 committed:
feat : Network 의존성 주입을 위한 프로토콜 생성, Mock 데이터 구현

- Network 의존성 주입을 위한 URLSessionProtocol, URLSessionDataTaskProtocol 프로토콜 생성 및 채택
- Network 테스트를 위한 DummyData 생성
- Network 테스트를 위한 MockURLSession, MockURLSessionDataTask 생성

Gundy93 committed:
feat: API 통신을 담당하는 NetworkManager 구현

- 기존 코드 서로 호환되게 리팩토링

wonbi92 committed:
refactor : decode메서드 로직 수정, 테스트를 위한 MockData, DataLoader 생성, 폴더 정리

- JSONDecoder의 decode메서드의 매개변수 타입을 String에서 Data로 수정
- JSONDecoder의 decode메서드가 디코딩중 생기는 에러에 맞는 문구를 출력하도록 수정
- 테스트를 위한 파일인 products.json파일을 테스트폴더로 이동
- 프로젝트 폴더에있는 데이터를 가져오는 DataLoader 구현
- 폴더 계층구조 정리

Gundy93 committed:
feat: NetworkManager를 검증하기 위한 NetworkManagerTests 생성

- OpenMarketTests 이름 JSONParsingTests으로 변경
- URLRequest를 매개변수로 받는 메서드들을 URL을 매개변수로 받도록 수정
- MockURLSessionDataTask의 init에 DummyData 파라미터 추가하여 초기화

wonbi92 committed:
test : NetworkManagerTests의 테스트 케이스 추가

- 네이밍 리팩토링 진행
- productListRequest와 dummyData의 인스턴스가 테스트 클래스의 프로퍼티로 존재하도록 리팩토링
- 테스트 케이스 추가
- ProductListRequest의 queryParameters의 값을 초기화할때 넘겨받아 추가하도록 리팩토링
- NetworkError에 failRequest케이스 추가

Gundy93 committed:
test : NetworkRequest를 테스트하는 NetworkRequestTests 파일 생성

- HealthCheckerRequest를 테스트하는 케이스 작성

wonbi92 committed:
test : ProductListRequest에 대한 테스트 케이스 추가

- NetworkRequestTests 테스트 케이스 네이밍 수정

Gundy93 committed:
test : ProductDetailRequest를 검증하는 테스트 케이스 추가

- urlComponents 이름 url로 변경
- queryParameters가 없을 경우 url 끝에 물음표가 붙지않도록 로직 리팩토링

wonbi92 committed:
refactor : 실제 서버에 있는 데이터와 일치하도록 Mock데이터 수정

- ProductList의 CodingKeys삭제
- dateFormatter의 dateFormat변경, timeZone 설정
- product Mock데이터 실제 서버와 일치하도록 수정

Gundy93 committed:
test: test code 수정

wonbi92 committed:
docs : README 추가

wonhee009 commented on OpenMarket/Network/HttpMethod.swift:
```diff
@@ -0,0 +1,29 @@
+//
+
+enum HttpMethod {
+    case get
+    case post
+    case put
+    case patch
+    case delete
```
해당 메서드를 모두 사용하나요?

wonbi92 replied:
step1에서는 get만 사용을 하긴 하지만, 이후 스탭에서 post, patch, delete등의 다른 메서드를 사용하기도 하고, httpMethod의 종류를 공부해보자는 의미로 주요한 메서드들은 구현해보자 라고 결론을 내려서 5개의 주요한 메서드 케이스를 모두 작성해 보았습니다!

wonhee009 replied:
post, patch, delete에 대해서는 이후 스텝에서 필요한 사항이라 구현했다고 이해했습니다.
그럼 put은 제외하는건 어떨까요?
"httpMethod의 종류를 공부해보자는 의미로 주요한 메서드들은 구현해보자" 이 부분은 `case put`과는 관계 없다고 생각 듭니다.
이렇게 모든 httpMethod가 열려 있으면 이후 다른 개발자가 봤을때 put에 대한 API 구현이 있다 생각할 수 있을거 같고, 이런 부분들이 이후 에러를 발생시킬 수 있을거 같습니다.
이 부분은 이후 스텝에서 확인해주셔도 좋습니다:)

wonhee009 commented on OpenMarket/OpenMarketTests/MockData/DataLoader.swift:
```diff
@@ -0,0 +1,33 @@
+
+import Foundation
+
+final class DataLoader {
```
static func만 갖고 있는 클래스로 보이는데요.
클래스로 정의한 이유가 있을까요?

wonbi92 replied:
DataLoader는 로컬에 있는 데이터를 Data타입으로 로드시켜주는 객체입니다. 로컬에 있는 파일의 URL을 만들어주는 메서드 안에 Bundle(for:)메서드에서 매개변수로 self를 넣어주는데, 이 과정에서 매개변수 타입이 클래스만 가능한 AnyClass이기 때문에 class로 정의하게 되었습니다.

wonhee009 commented on OpenMarket/OpenMarket/Network/OpenMarketNetworkRequest.swift:
```diff
@@ -0,0 +1,39 @@
+    let httpMethod: HttpMethod = .get
+    let urlHost: String = "https://openmarket.yagom-academy.kr/"
+    let urlPath: String = "api/products"
+    var queryParameters: [String : String] = [:]
+    
+    init(pageNo: Int, itemsPerPage: Int, searchValue: String = "") {
+        self.queryParameters["page_no"] = String(pageNo)
+        self.queryParameters["items_per_page"] = String(itemsPerPage)
```
```suggestion
    var queryParameters: [String : String]
    
    init(pageNo: Int, itemsPerPage: Int, searchValue: String = "") {
        queryParameters[
           "page_no": String(pageNo),
           "items_per_page": String(itemsPerPage)
```
이렇게도 가능할거 같습니다.

wonbi92 replied:
오 이렇게도 가능할 것 같습니다! 조언주신 방법으로 수정 해보았습니다!

wonhee009 commented on OpenMarket/OpenMarket/Network/DTO/ProductList.swift:
```diff
@@ -0,0 +1,20 @@
+
+import Foundation
+
+struct ProductList: Codable {
```
Model을 정의한 파일들이 DTO라는 폴더 안에 위치한거로 보이는데요.
DTO보다는 Model이라는 폴더명은 어떤가요?

wonbi92 replied:
코맨트 주신 부분을 토대로 Gundy와 상의해본 결과, Model로 바꾸는것이 코드를 이해하는 데 더 나은 방법이라는 결론이 나왔습니다. 폴더명을 Model로 변경하였습니다!

wonhee009 commented on OpenMarket/OpenMarket/ViewController.swift:
```diff
@@ -10,9 +10,39 @@ class ViewController: UIViewController {
     override func viewDidLoad() {
         super.viewDidLoad()
-        // Do any additional setup after loading the view.
+        let networkManager = NetworkManager()
+        let healthCheckerRequest = HealthCheckerRequest()
+        let productListRequest = ProductListRequest(pageNo: 1, itemsPerPage: 20)
+        let detailRequest = ProductDetailRequest(productID: 183)
+        
+        guard let healthCheckerURL = healthCheckerRequest.url,
+              let productListURL = productListRequest.url,
+              let detailURL = detailRequest.url else { return }
+        
+        networkManager.fetchData(for: healthCheckerURL, dataType: String.self) { respone in
+            switch respone {
+            case .success(let data):
+                print(data)
+            case .failure(let error):
+                print(error)
+            }
+        }
+        networkManager.fetchData(for: productListURL, dataType: ProductList.self) { respone in
+            switch respone {
+            case .success(let data):
+                print(data)
+            case .failure(let error):
+                print(error)
+            }
+        }
+        networkManager.fetchData(for: detailURL, dataType: Product.self) { respone in
+            switch respone {
+            case .success(let data):
+                print(data)
+            case .failure(let error):
+                print(error)
```
테스트 해보기 위한 코드일까요?
현재 스텝 요구사항에는 없는 부분인거 같아 질문드립니다.

Gundy93 replied:
네트워킹에 대한 테스트 코드가 맞습니다! 현재는 지웠습니다!

wonhee009 commented on OpenMarket/Network/NetworkRequest.swift:
```diff
@@ -0,0 +1,38 @@
+}
+
+extension NetworkRequest {
+    var url: URL? {
+        if queryParameters.isEmpty {
+            return URL(string: urlHost + urlPath)
+        }
+        var urlComponents = URLComponents(string: urlHost + urlPath)
+        let queryItems = queryParameters.map { URLQueryItem(name: $0, value: $1) }
+        urlComponents?.queryItems = queryItems
+        
+        return urlComponents?.url
+    }
```
해당 프로퍼티는 외부에서 접근할 일이 있나요?

Gundy93 replied:
URL을 매개변수로 fetchData를 구현했기 때문에 테스트 과정에서 url이라는 프로퍼티에 접근하여 사용했습니다. 해당 프로퍼티를 private 처리하고, URLRequest 객체에서 url 프로퍼티에 접근하는 방식으로 로직을 수정하였습니다.

wonhee009 commented on OpenMarket/Extension/String+.swift:
```diff
@@ -0,0 +1,24 @@
+        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss"
+        formatter.timeZone = TimeZone(abbreviation: "UTC")
+        return formatter
+    }
```
모든 String 타입에서 해당 프로퍼티에 접근가능할거 같은데요.
해당 프로퍼티는 한정된 곳에서만 사용될거 같은데 어떻게 생각하시나요?

Gundy93 replied:
아래 내용과 연결되는 부분 같아서 아래에 더 자세히 적었습니다! 해당 부분을 String Extension에서 제거하였습니다.

wonhee009 commented on OpenMarket/Extension/String+.swift:
```diff
@@ -0,0 +1,24 @@
+            return result
+        }
+        return nil
+    }
```
해당 메서드도 동일하게 모든 String 타입에서 접근가능할거 같은데요.
모든 String 타입에서 접근 가능한 방식 말고 DateConverter(예시) 같은 하나의 class/struct를 추가하는건 어떤가요?

Gundy93 replied:
String Extension을 삭제하고, 이후 `price` 및 다른 가격에 대해서도 numberFormatter를 사용하게 될 것 같아 FormatConverter라는 구조체를 구현하였습니다.

wonhee009 commented on OpenMarket/Extension/JSONDecoder+.swift:
```diff
@@ -0,0 +1,30 @@
+        var decodedData: T?
+        do {
+            decodedData = try decoder.decode(type, from: data)
+        } catch DecodingError.keyNotFound(let key, let context) {
+            print("could not find key \(key) in JSON: \(context.debugDescription)")
+        } catch DecodingError.valueNotFound(let type, let context) {
+            print("could not find type \(type) in JSON: \(context.debugDescription)")
+        } catch DecodingError.typeMismatch(let type, let context) {
+            print("type mismatch for type \(type) in JSON: \(context.debugDescription)")
+        } catch DecodingError.dataCorrupted(let context) {
+            print("data found to be corrupted in JSON: \(context.debugDescription)")
```
catch 문에서 다른 부분이 출력 메시지가 다른거로 보이는데요.
[localizedError](https://developer.apple.com/documentation/foundation/localizederror)의 errorDescription을 이용해 보는건 어떤가요?

wonbi92 replied:
코맨트 주신 부분을 정확하게 이해하지 못했지만, error의 localizedDescription 프로퍼티를 활용하게 바꿔보았습니다.
```Swift
extension DecodingError {
    public var errorDescription: String {
        switch self {
        case .typeMismatch(let type, let context):
            return "type mismatch for type \(type) in JSON: \(context.debugDescription)"
        case .valueNotFound(let type, let context):
            return "could not find type \(type) in JSON: \(context.debugDescription)"
        case .keyNotFound(let key, let context):
            return "could not find key \(key) in JSON: \(context.debugDescription)"
        case .dataCorrupted(let context):
            return "data found to be corrupted in JSON: \(context.debugDescription)"
        @unknown default:
            return localizedDescription
        }
    }
}

do {
    decodedData = try decoder.decode(type, from: data)
} catch let error as DecodingError {
    print(error.errorDescription)
} catch {
    print(error.localizedDescription)
}
```
혹시 라자냐가 코맨트 주신 부분의 방향성이 위 코드와 같은 걸까요? 저희가 올바르게 이해했는지 잘 모르겠습니다 ㅠㅠ

wonhee009 replied:
LocalizedError를 확인해보면 프로토콜인데요.
```swift
enum DecodingError: LocalizedError {
   case .typeMismatch
   ...

   var errorDescription: String {
      switch self {
         case .typeMismatch(): ...
      }
   }
}
```
이런 방식으로 생각했습니다.

Gundy93 replied:
이미 `DecodingError`는 LocalizedError 프로토콜을 채택한 에러인데, `errorDescription`은 nil을 반환하더라고요! extension을 통해 nil이 아닌 적절한 String을 반환하게끔 만들어본거였습니다. 😃

wonhee009 reviewed (changes requested):
수고하셨습니다:)
코멘트 남겨두었으니 확인 부탁드려요!

- 폴더링은 팀마다도 룰이 다르고 개인 취향도 있어서 뭐가 맞다고 말씀 드릴 순 없을거 같습니다.
그래도 각자의 역할에 맞게 분리되면 될거 같습니다.
- 넵 그냥 이어나가면 될거 같습니다.

wonbi92 committed:
refactor : OpenMarketNetworkRequest 로직 및 폴더명 변경

- urlHost, urlPath 의 /값 위치 변경
- queryParameters를 초기화 하는 방식 변경
- DTO폴더명을 Model로 변경

Gundy93 committed:
refactor : String+ 삭제 및 FormatConverter 구현, 접근제어자 수정

- String+를 삭제하고 DateFormatter 및 이후 NumberFormatter의 기능을 수행할 FormatConverter 구조체 구현
- NetworkRequest의 프로퍼티 url을 private로 접근제어자 수정, requestURL 네이밍 request로 변경
- JSON 파싱할 타입에 formatConverter 프로퍼티 추가
- viewDidLoad 내부 테스트 코드 삭제
- 테스트 코드 로직 수정

wonbi92 committed:
refactor : JSONDecoder의 decode 로직 수정

- error처리시 error.localizedDescription를 출력하도록 로직 수정

Gundy93 committed:
refactor : DecodingError extension, HttpMethod 수정

- extension을 통한 DecodingError의 errorDescription 프로퍼티 값 수정
- HttpMethod에서 put 삭제

wonhee009 reviewed (approved):
수고하셨습니다 :)
